### ISSUE TITLE
Missing Content-Encoding

### ISSUE THREAD

**Issue by den-crane** [ClickHouse/ClickHouse#64802]
Clickhouse-server HTTP server does not send `Content-Encoding` in some cases starting from 24.3
It breaks clients libraries, python, java. Because they rely on `Content-Encoding` to decode HTTP stream.

It happens only with long queries which return a lot of data and only when `send_progress_in_http_headers=1`.
I used `sleep` to emulate long lasting queries. 

sleep(.001)
```
$ (curl -Ss -vvvv -o outx1  -H 'Accept-Encoding: zstd' "localhost:8123?send_progress_in_http_headers=1&enable_http_compression=1&wait_end_of_query=0&query=select+distinct+sleep(.001),name+from+generateRandom('name+String',1,1000,2)+limit+100009+format+TSV" 2>&1) |grep 'Content-Encoding'
```
curl output: **[ < Content-Encoding: zstd ]** -- expected output, everything is good.

sleep(.1)
```
$ (curl -Ss -vvvv -o outx2 -H 'Accept-Encoding: zstd' "localhost:8123?send_progress_in_http_headers=1&enable_http_compression=1&wait_end_of_query=0&query=select+distinct+sleep(.1),name+from+generateRandom('name+String',1,1000,2)+limit+100009+format+TSV" 2>&1) |grep 'Content-Encoding'
```
curl output: **[] (no Content-Encoding key in HTTP header)** -- not expected output,  expected Content-Encoding: zstd

data is correct, only HTTP header is broken.
```
ls -lh outx1 outx2
-rw-r--r-- 1 denis.zhuravlev WebAMG.StJohn 41M Jun  4 17:20 outx1
-rw-r--r-- 1 denis.zhuravlev WebAMG.StJohn 41M Jun  4 17:20 outx2

file  outx*
outx1: Zstandard compressed data (v0.8+), Dictionary ID: None
outx2: Zstandard compressed data (v0.8+), Dictionary ID: None
```

https://github.com/ClickHouse/clickhouse-connect/issues/347

related to `send_progress_in_http_headers`

**Comment by den-crane** [comment]
@alexey-milovidov  ?

**Comment by pkit** [comment]
Looks like a bug to me, can be fixed easily though:

```diff
diff --git a/src/Server/HTTP/WriteBufferFromHTTPServerResponse.cpp b/src/Server/HTTP/WriteBufferFromHTTPServerResponse.cpp
index 8098671a903..3dd8c67fc0a 100644
--- a/src/Server/HTTP/WriteBufferFromHTTPServerResponse.cpp
+++ b/src/Server/HTTP/WriteBufferFromHTTPServerResponse.cpp
@@ -143,6 +143,10 @@ void WriteBufferFromHTTPServerResponse::onProgress(const Progress & progress)
         accumulated_progress.incrementElapsedNs(progress_watch.elapsed());
         progress_watch.restart();
 
+        /// We do not really care about `initialized` here, as this one just sets a value in a map
+        if (compression_method != CompressionMethod::None)
+            response.set("Content-Encoding", toContentEncodingName(compression_method));
+
         /// Send all common headers before our special progress headers.
         startSendHeaders();
         writeHeaderProgress();

```

**Comment by alexey-milovidov** [comment]
@pkit, looks like a good fix, please send a pull request.

**Comment by guillaumematheron** [comment]
We're seeing this issue on a server where `send_progress_in_http_headers = 0`

> ClickHouse server version 24.8.4.13 (official build).

Content-Encoding is absent, causing a parsing error in the python client. Disabling compression in `get_client` prevents the error.

**Comment by Algunenano** [comment]
https://github.com/ClickHouse/ClickHouse/pull/68975

**Comment by den-crane** [comment]
@guillaumematheron 
1. there is stable releases with a fix.
2. send_progress_in_http_headers usually enabled by a client. See https://github.com/ClickHouse/clickhouse-connect/issues/122#issuecomment-1426226111 `get_client(send_progress=False)`

**Comment by siradjev** [comment]
Is there a chance to back-port the fix to the 24.3-lts?

**Comment by thevar1able** [comment]
@siradjev https://github.com/ClickHouse/ClickHouse/pull/71482
